### PR TITLE
MDEV-10814: hartmut coredump exclusions (10.1)

### PR DESCRIPTION
**Pull request by grooverdan**
re-based previous work. one minor simplification.
TODO: MYISAM_KEY_BUFFER

### CODE REVIEW

**Comment by svoj** [comment]
Hi Daniel,

Thanks for your contribution. JIRA task just for the record: https://jira.mariadb.org/browse/MDEV-10814

This task was added to 10.1 backlog, which is planned to be handled between 2017-04-29 and 2017-05-05.

Thanks,
Sergey

File: include/my_core_no_dump.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+   when providing a core dump to a 3rd party for analysis
+
+ * security: certain buffers, especially the InnoDB pool buffer or the
+   Aria page cache, are likely to contain sensitive user data. Excluding
+   these from a core dump can improve data security and especially can be
+   a requirement for being able to pass production server core dumps to
+   3rd parties for analysis
+*/
+
+#include <my_global.h>
```
Comment (svoj) [inline include/my_core_no_dump.h]: Do we really need to include my_global.h if HAVE_MADV_DONTDUMP is not defined?

File: sql/sys_vars.cc [hunk 2]
```diff
@@ -63,6 +63,8 @@
 #include "opt_range.h"
 #include "rpl_parallel.h"
 
+#include "my_core_no_dump.h"
```
Comment (svoj) [inline sql/sys_vars.cc]: Do we need this include here?

File: Docs/README-core-dump-exclusion [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+can't be changed dynamically at run time as exclusions happen right
+after allocating memory at startup.
+
+Possible values for "core_nodump" at this point are:
+
+ Value              | Description                                          
+--------------------|------------------------------------------------------
+ NONE               | Default: keep current behavior that dumps everything 
+ INNODB_POOL_BUFFER | Exclude InnoDB pool buffer instances from core dump  
+ MYISAM_KEY_BUFFER  | Exclude MyISAM key buffer(s) from core dump
```
Comment (svoj) [inline Docs/README-core-dump-exclusion]: MYISAM_KEY_BUFFER is not available, why?

File: storage/xtradb/buf/buf0buf.cc [hunk 4]
```diff
@@ -65,6 +65,9 @@ Created 11/5/1995 Heikki Tuuri
 #include "fil0pagecompress.h"
 #include "ha_prototypes.h"
 
+/* for core dump specific stuff */
```
Comment (svoj) [inline storage/xtradb/buf/buf0buf.cc]: Pff... such comments make little to none sense.

File: include/my_core_no_dump.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+#ifdef HAVE_MADV_DONTDUMP
+
+#include <sys/mman.h>
+
+/* Core dump exclusion constants */
+#define CORE_NODUMP_NONE                (0)
+#define CORE_NODUMP_MAX                 (1 << 1)
+#define CORE_NODUMP_INNODB_POOL_BUFFER  (1 << 2)
```
Comment (svoj) [inline include/my_core_no_dump.h]: Why POOL_BUFFER versus commonly used BUFFER_POOL?

File: Docs/README-core-dump-exclusion [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+(MADV_DONTDUMP) starting with kernel version. Such memory regions need to
+be page aligned so out of the box the feature is only useful when dealing
+with large enough malloc() allocations or when directly dealing with
+mmap()ed or shared memory regions directly.
+
+Certain memory allocations now call the new function
+
+  exclude_from_coredump(void *ptr, size_t size, ulonglong flagmask);
+
+after successfully aquiring memory. A global variable "core_nodump"
```
Comment (svoj) [inline Docs/README-core-dump-exclusion]: acquiring

File: include/my_core_no_dump.h [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+extern ulong opt_core_nodump;
+
+inline void exclude_from_coredump(void *ptr, size_t size, ulong flags)
+{
+  if (opt_core_nodump & (flags | CORE_NODUMP_MAX)) {
```
Comment (svoj) [inline include/my_core_no_dump.h]: I'd prefer to move ifdef here, so that you don't have to have \#else part

File: include/my_core_no_dump.h [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA */
+
+/* This is the include file that should be included 'first' in every C file. */
+
+#ifndef MY_CORE_NO_DUMP_INCLUDED
+#define MY_CORE_NO_DUMP_INCLUDED
+
+/*
+ Core dump control options to exclude certain buffers from core dump files
```
Comment (svoj) [inline include/my_core_no_dump.h]: Comment indentation doesn't follow coding style: should be 2 spaces.

File: include/my_core_no_dump.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+#include <sys/mman.h>
+
+/* Core dump exclusion constants */
+#define CORE_NODUMP_NONE                (0)
+#define CORE_NODUMP_MAX                 (1 << 1)
+#define CORE_NODUMP_INNODB_POOL_BUFFER  (1 << 2)
+
+extern ulong opt_core_nodump;
+
+inline void exclude_from_coredump(void *ptr, size_t size, ulong flags)
```
Comment (svoj) [inline include/my_core_no_dump.h]: static inline please.

**Comment by svoj** [comment]
@serg, do you think we need Docs/README-core-dump-exclusion? Probably leave it for KB? Should we also have core_nodump sysvar when it is not supported, so that we can still start the server if it was specified? Should exclude_from_coredump() be a service?

@dr-m, @janlindstrom could you confirm that buffer pool pages once allocated are never freed or reallocated till InnoDB shutdown?

@grooverdan, nice feature, but I guess this adds another glitch that'll make UNINSTALL PLUGIN innodb keep these memory areas not dumpable.

**Comment by vuvova** [comment]
yeah, no need to add more files to `Docs/` — nobody looks there anyway. KB is a better place for it.

Service? Yes, as InnoDB uses it.

But also, I don't like "INNODB_POOL_BUFFER" in the server code. There are few ways to fix it:
* have `@@innodb_nodump_pool_buffer` sysvar, instead of `@@core_nodump`. This will becomes inconvenient when many engines will have their own no-dump regions.
* have `@@core_nodump` and every engine can dynamically add new values to the set of no-dump regions. That's too complicated, at least for now.
* Just drop `@@core_nodump` and don't dump innodb pool buffer unconditionally. I'd prefer that.

**Comment by svoj** [comment]
I believe buffer pool dump might be useful for debugging corruptions.

If we move system variable into InnoDB, implementing service might be overkill. I'd rather use madvise directly.

**Comment by vuvova** [comment]
Agree with using madvise directly

I think unconditional is ok, was this buffer pool in the core dump ever useful? At least once? But I don't have a strong opinion about it, variable is ok too. Just not in the server.

**Comment by grooverdan** [comment]
The goal with this was to reduce the size to a point where collecting coredumps is quick and low risk. With large systems with many G of buffer pool the assurance of space available on disk for a core dump wasn't so certain. I've never needed to look at the innodb buffer pool in a core dump mainly because the variables of the crashed program have contained the necessary information (probably due to innodb stablility especially in the pool area). Happy to keep this as a variable however.

I'm happy to do this as innodb_nodump_pool_buffer rather than a service.

The dynamic allocation of myisam key buffer was the reason @hholzgra didn't get to it and me neither. In many ways some of these madvise/THP and other aspects should be part of jemalloc.

10.2 will need to account for the online resizing of buffer pools and I will need check what actions are taken on UNINSTALL PLUGIN, probably a reversing madvice call to be sure that the same area doesn't get reallocated by jemalloc. As such I'll just retarget to 10.2 that rather than 10.1.

**Comment by svoj** [comment]
Summary:
* keep cmake stuff in the server
* use madvise() directly
* no system variables in first patch (we may have to add them in future though)
* "unadvise" whenever memory hunk is freed
* 10.2 is safer choice
* no doc, better add it to jira for Ian's reference

**Comment by hholzgra** [comment]
On 20.04.2017 01:00, Daniel Black wrote:
> The goal with this was to reduce the size to a point where collecting
> coredumps is quick and low risk. With large systems with many G of
> buffer pool the assurance of space available on disk for a core dump
> wasn't so certain.

disk size *and* time it takes to write a full core dump to disk were
original concerns that came up in some conversation with booking.com
back then ... writing back data in the order of 100G can delay process
restart (by mysqld_safe, systemd, or whatever) quite a bit and so
substantially affect MTTR (mean time to recover)

Another aspect, especially that we have data-at-rest encryption now,
would be to exclude the buffer pool and some other buffers for
security reasons.

What use would an encrypted tablespace be if large parts of it could
still be seen in unencrypted form in a core dump file?

> I've never needed to look at the innodb buffer pool
> in a core dump mainly because the variables of the crashed program have
> contained the necessary information (probably due to innodb stablility
> especially in the pool area). 

I think I had only one such case in all those years, in 4.1 or 5.0 days
in an issue escalated to Heikki back then, where looking into actual
buffer pool page contents was needed ... at that rate we wouldn't loose
much with it ...

> In many ways some of these madvise/THP and other aspects should be
part of jemalloc.

Yes, I'd love that, too. As NODUMP is a page level attribute, not a byte
level or byte range one, it may not be high on their list though, as it
would either need a specific NODUMP pool to allocate from, or would
require full page allocations for small malloc() requests, leaving
the rest of the page empty.

**Comment by grooverdan** [comment]
rework provided. Thanks for the feedback. #364 preferred.

**Pull request closed without merging**